ApprovalTests/integrations/CheckFileMacroIsAbsolute.h

```diff
@@ -0,0 +1,19 @@
+#ifndef APPROVALTESTS_CPP_CHECKFILEMACROISABSOLUTE_H
+#define APPROVALTESTS_CPP_CHECKFILEMACROISABSOLUTE_H
+
+// Maintenance note: the following help message must be short, and
+// not contain any newline characters, as their display looks messy
+// in some compiler outputs.
+
+#ifdef _MSC_VER
```

Comment:
Good point. I'll check into it some more...

(I think tested it on Cygwin, but not MSYS.)

This looks useful: https://blog.kowalczyk.info/article/j/guide-to-predefined-macros-in-c-compilers-gcc-clang-msvc-etc..html

We must have got it working in MSYS before, as we have: ApprovalTests/utilities/WinMinGWUtils.h

I think that most people people use ApprovalTests via the single-header download. This is based on the fact that lots of people have used it, but we have had very few people asking for it to play more nicely with CMake...

Checks in CMake won't help users of the single-header... Their first line of defence will be any static asserts, and the second will be the run-time checks.

So an advantage of compile-time checks in code is that they will work whatever build system people use...

This makes me wonder if it would be OK for now to focus on getting the static_asserts right - like for MSYS - and not make the CMake more complicated for Ninja?

Reply:
Duh, about the single-header usage. I overlooked that. 😓 The CMake script might be worth it at some point, but it really is more a nicety for those using CMake.

My concern is the [MSYS Makefiles](https://cmake.org/cmake/help/latest/generator/MSYS%20Makefiles.html) CMake generator which I don't know how to detect at compile-time and uses weird, Unix-like absolute paths, but it does say that it uses make so it may not have any problems.

I still think the ```_WIN32``` macro is a safer bet because it should work for anything on Windows. Of course, I hope Cygwin doesn't define that macro because it uses Unix style paths, I think.

That said, we can probably just let the corner cases crop-up and fix them ad-hoc.

FYI, [Boost Predef](https://www.boost.org/doc/libs/1_70_0/doc/html/predef/reference.html) is a fantastic reference for various system and compiler macros.

Reply:
@jwillikers just a quick note to say thanks very much for all your input and links on this - it's really helpful, and I will pick it all up soon...

Reply:
Your welcome! 😊

Reply:
@ jwillikers, an update:

We now have a [CI build for mingw and Cygwin that passes](https://ci.appveyor.com/project/isidore/approvaltests-cpp/builds/30061900) and that uses the MSYS Makefiles generator.

> Should this be something more generic for Windows not just the MSVC compiler?
> This would be my thought `#ifdef _WIN32`.

Yes, you were right that I needed to [change the ifdef to `_WIN32`](https://github.com/approvals/ApprovalTests.cpp/commit/4486afea0544d4a20b71f1b8d3a6895d50c1118e) - thank you.

> There's a layer of complexity here when MSYS is being used because it allows absolute paths like Windows and paths like `/c/path/to/file`. I can not find a good way to detect MSYS in the header file, which means we should probably use CMake to check this.

I started looking at reproducing that, and eventually decided that as I've now got a passing build with MSYS Makefiles generator, it was better to spend time on main remaining Ninja actions for now. If anyone does report problems with MSYS paths, I'll revisit this.

For future reference, we do already detect mingw builds in [WinMinGWUtils.h](https://github.com/approvals/ApprovalTests.cpp/blob/4ff9cfbe8dbd4de6ae4aebe5fd6b7272c54aed0a/ApprovalTests/utilities/WinMinGWUtils.h#L6) with:

```c++
#if (defined(__MINGW32__) || defined(__MINGW64__))
```

I'll review the remaining definition-of-done actions for this PR and then merge it later this morning, so I can press on with documenting the Ninja problems better, and then do a point release, probably in the next day or two.